lib/bolt/plugin/puppetdb.rb

```diff
@@ -15,9 +16,77 @@ def hooks
         ['lookup_targets']
       end
 
+      def warn_missing_fact(certname, fact)
+        @logger.warn("Could not find fact #{fact} for node #{certname}")
+      end
+
+      def fact_name(raw_fact)
+        fact = raw_fact.split(".")
+        fact.drop(1) if fact[0] == 'facts'
+      end
+
       def lookup_targets(opts)
-        nodes = @puppetdb_client.query_certnames(opts['query'])
-        nodes.map { |certname| { 'uri' => certname } }
+        targets = @puppetdb_client.query_certnames(opts['query'])
+        facts = []
+        facts << fact_name(opts['name']) if opts.key?('name')
```

Comment:
We had discussed not making config special but allowing all of the "target data" keys here. I think a generic solution will be easier than treating config special.

```ruby
target_opts = opts.select {|k, _| ['uri', 'name', 'config', (maybe other options)].includes?(k) }
Bolt::Util.walk_vals(target_opts) do |value|
  ...
end
```

Reply:
Should we also apply this pattern to the terraform plugin? This PR is very much based on that - I can put up a separate pr as part of this ticket for that?

Reply:
I don't think we have anything analagous to $facts with the terraform plugin.